test/gradualizer_dynamic_suite.erl

```diff
@@ -0,0 +1,60 @@
+-module(gradualizer_dynamic_suite).
+
+-export([reload/1]).
+
+-include_lib("common_test/include/ct.hrl").
+-include_lib("stdlib/include/assert.hrl").
+
+reload(Config) ->
+    Module = ?config(dynamic_suite_module, Config),
+    ?assert(Module /= undefined),
+    case erlang:function_exported(Module, generated_tests, 0) of
+        true ->
+            {ok, Module:generated_tests()};
```

Comment:
What's this test suite doing?

It looks like a lot of magic. I don't like magic. :)

Reply:
There's a little bit of magic, but let's start from the standard CT conventions. CT treats files matching `*_SUITE.erl` as test suites, so `gradualizer_dynamic_suite.erl` is not a test suite, it's a helper file for use in test suites. The actual test suites match the previous tests defined in EUnit:
- `test/known_problems_should_fail_SUITE.erl`
- `test/known_problems_should_pass_SUITE.erl`
- `test/should_fail_SUITE.erl`
- `test/should_pass_SUITE.erl`.

The dynamic suite helper exports only one function - `gradualizer_dynamic_suite:reload/1` - which generates CT test cases based on passed in config and then dynamically reloads the module it's invoked from. It's idempotent, so calling it once, twice, or 100 times has the same effect. That's where the name comes from - "dynamic suite". The reason for this helper is that, unlike EUnit, CT has no standard way to generate tests dynamically.

The tests are generated for each of the test files defined under the respective should pass/fail/known problems directories.

The most enigmatic part is where to actually invoke `gradualizer_dynamic_suite:reload/1` in a client test suite - this is something I had to dive in OTP CT code to figure out. Apparently, `*_SUITE:groups/0` is the first function CT calls when running a test suite, so it's the best place to call `reload/1` from. Before @xxdavid's remark that listing cases by hand is tedious we could call `reload/1` from something more logical, like `init_per_suite/1`, but if we want to avoid listing the to-be-generated tests manually, we have to play with how CT runs the test suites.